packages/cosmic-swingset/lib/ag-solo/vats/types.js

```diff
@@ -1,7 +1,35 @@
+// @ts-check
+
 /**
  * @typedef {Object} Board
  * @property {(id: string) => any} getValue
  * @property {(value: any) => string} getId
  * @property {(value: any) => boolean} has
  * @property {() => string[]} ids
  */
+
+/**
+ * @typedef {Object} NameHub
+ * @property {(...path: Array<unknown>) => Promise<unknown>} lookup Look up a
+ * path of keys starting from the current NameHub.  Wait on any reserved
+ * promises.
+ */
+
+/**
+ * @typedef {Object} NameAdmin
+ * @property {(key: unknown) => void} reserve Mark a key as reserved; will
+ * return a promise that is fulfilled when the key is updated (or rejected when
+ * deleted).
+ * @property {(key: unknown, newValue: unknown) => void} update Fulfill an
+ * outstanding reserved promise (if any) to the newValue and set the key to the
+ * newValue.
```

Comment:
I was wondering why `reserve` was a separate function, and I think I know why, but let me check:

Is it that `reserve` is allowing the NameAdmin to say - I will name this, but in a sec, and you can wait for it? And if you try to look up a name I haven't reserved, you don't have to wait for me to name it, I'll just tell you it doesn't have a name? 
In other words, if there was just one function (`add`), we might have users waiting for a response forever, not knowing the NameAdmin has no intention of naming it, so this is solving that problem?

Reply:
Yes. We want them to be initialization-order insensitive for names that will be supplied, but fail promptly if they are requesting a name that will not be provided.

Reply:
Yes, that's exactly right.  It was @dtribble's design, to give better errors but still allow for the economy bootstrapping to be done without subtle order dependencies (IIUC, especially wrt MOE issuers vs the AMM).